enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,924 @@
+in a namespace with a user-defined primary network will also be limited to only accessing the cluster IP of the services
+for that network it. Load balancer IP and nodeport services are also supported on user-defined networks. Service
+selectors are only able select endpoints from the same namespace where the service exists. Services that exist before
+the user-defined network is assigned to a namespace will result in OVN-Kubernetes executing a re-sync on all services in
+that namespace, and updating all load balancers. Keep in mind that pods must not exist in the namespace when the
+namespace is assigned to a new network or the new network assignment will not be accepted by OVN-Kubernetes.
+
+Services in a user-defined network will be reachable by other namespaces that share the same network.
+
+As previously mentioned, Kubernetes API and DNS services will be accessible by all pods.
+
+Endpoint slices will provide the IPs of the cluster default network in Kubernetes API. For this implementation the required
+endpoints are those IP addresses which reside on the user-defined primary network. In order to solve this problem,
+OVN-Kubernetes may create its own endpoint slices or may choose to do dynamic lookups at runtime to map endpoints to
+their primary IP address. Leveraging a second set of endpoint slices will be the preferred method, as it creates less
+indirection and gives explicit Kube API access to what IP addresses are being used by OVN-Kubernetes.
+
+Kubelet health checks to pods are queried via the cluster default network. When endpoints are considered unhealthy they
+will be removed from the endpoint slice, and thus their primary IP will be removed from the OVN load balancer. However,
+it is important to note that the healthcheck is being performed via the default interface on the pod, which ensures the
+application is alive, but does not confirm network connectivity of the primary interface. Therefore, there could be a
+situation where OVN networking on the primary interface is broken, but the default interface continues to work and
+reports 200 OK to Kubelet, thus rendering the pod serving in the endpoint slice, but unable to function. Although this
+is an unlikely scenario, it is good to document.
+
+### Network Policy
+
+Configuring network policies that allow traffic between user-defined network namespaces will not be invalidated or
+ignored by OVN-Kubernetes. If a user configures a network policy that would have no effect, (like stopping namespace A
+from talking to B on different networks) then OVN-Kubernetes will still configure this policy in OVN. These policies
+will also have no effect between namespaces that are assigned to different networks as there will be no network
+connectivity between the namespaces. However, there may still be a desire to configure network policy between namespaces
+that share the same network, as well as for other reasons like ingress/egress control.
+
```

Comment:
ANP will have no ability to allow traffic between two networks. However, any network policy that affects ingress/egress cluster traffic will still be enforced. Same with egress firewall.

I added a sentence for this.

Any policy will be enforced on LSP that belongs to the user-defined network.

Reply:
@trozet : So "ANP will have no ability to allow traffic between two networks." That is fine if the NetworkCRD is created by the admin, BUT if the NetworkCRD is created by the "user" or "tenant" then ANP has to take precedence because it was decided by the admin that nsA should be able to talk to nsB irrespective of how tenant thinks this should work? @danwinship wdyt?

Reply:
A user should understand the feature. If they use isolated networks, they are deciding to not allow communication between them. I don't see how this violates any tenant of ANP. By default, pods can communicate without restriction. ANP would be used to by default block traffic, and then allow holes. By using an isolated network, the user is not compromising security of ANP by allowing traffic that shouldn't be allowed by ANP. 

So if an admin creates an ANP that blocks all traffic between namespaces, except for one path. Two namespaces that are on different networks do not break this security, they still cant communicate for the blocked traffic. They also cannot communicate for the exceptional traffic. I dont see this is as a problem. The ANP at admin level should enforce what traffic should not be allowed, and what traffic may be OK to allow, but it doesnt need to guarantee at a tenant level that a tenant has to allow that traffic.

ANP will still be able to be enforced between two or more namespaces that share the same network.

Reply:
Admin with ANPs should be able to dictate what tenants can and cannot do. So if admin says nsA and nsB should be receiving ingress traffic from monitoring namespace always; and now if tenant of networkRed under which nsA and nsB fall under say they want isolation via multiple-networks so monitoring should not be able to scrape any data from them, that's wrong, admin wins here.

Reply:
> ANP would be used to by default block traffic, and then allow holes.

ANP can be used to forbid the user to block traffic themselves as well. (ie, "the monitoring namespace can access your pods, even if you say they can't")

But I agree that secondary networks are orthogonal to this. It's not about the admin "winning" or "losing", it's about what is and isn't in scope for ANP. ANP only allows controlling traffic that is part of the (standard) Kubernetes network model (well, plus anything that we decide to extend it to cover as part of this enhancement). 

Another example: a pod uses Multus to create a secondary interface, and then binds a socket to only listen on the IP of the secondary interface. In that case, the monitoring namespace can't reach that service via the pod network, and ANP has no ability to be able to force it to be able to connect via the secondary interface, right?

If the cluster admin wants _full_ control over networking, then they have to keep full control over networking. If they delegate some control to the namespace admins, then there will inherently be some things they can't do any more.

Reply:
ACK, I think saying ANP is out of scope for user-defined networks is fair enough OR the other way round where user-defined networks are not within the scope of ANP.  @trozet do you want to include that in your next update?

Reply:
> But I agree that secondary networks are orthogonal to this. It's not about the admin "winning" or "losing", it's about what is and isn't in scope for ANP. ANP only allows controlling traffic that is part of the (standard) Kubernetes network model (well, plus anything that we decide to extend it to cover as part of this enhancement).
> 

right so "winning" was probably not the right word here, what I meant was the "user personas" we defined with ANP and NetPol on how isolation works will either need to be told as "out of scope as far as user-defined networks are concerned" like you are suggesting OR we need to define that piece explicitly if we make the ANP controller aware of this.